### ISSUE TITLE
False-positive no-unused-expressions in do expression

### ISSUE THREAD

**Issue by PinkaminaDianePie** [babel/eslint-plugin-babel#149]
`no-unused-expressions` false positive in `do` expressions with `if` statements without braces.

```javascript
let foo = do {
  if (!a) 0;  // false positive here
  else if (b) 1;  // false positive here
  else if (c) 2;  // false positive here
  else 3; // false positive here
};
```
it works correctly only with braces:
```javascript
let foo = do {
  if (!a) { 0; }  // ok
  else if (b) { 1; }  // ok
  else if (c) { 2; }  // ok
  else { 3; } // ok
};
```
Second option is not good for me just because it conflitcs with `brace-style` from airbnb guide. First option is OK with airbnb, but it false-positive because of this bug.

**Comment by ljharb** [comment]
The second option is ok with the airbnb guide too.

**Comment by PinkaminaDianePie** [comment]
Maybe I'm doing something wrong, but for me airbnb guide shows `brace-style` error. My eslint config is little bit different, but without a lot of changes:

```javascript
module.exports = {
  "extends": "airbnb",
  "parser": "babel-eslint",
  "plugins": [
    "babel"
  ],
  "env": {
    "browser": true,
    "mocha": true
  },
  "rules": {
    "no-bitwise": "off",
    "no-restricted-syntax": ["error", "ForInStatement", "LabeledStatement", "WithStatement"],
    "max-depth": ["error", 4],
    "complexity": ["error", 11],
    "babel/no-unused-expressions": "error",
    "no-unused-expressions": "off",
  },
};

```

**Comment by ljharb** [comment]
Perhaps because it’s all in one line?

**Comment by PinkaminaDianePie** [comment]
maybe, but such style looks better with `do` expression, it looks a little bit similar to pattern matching - you have both condition and result expression in one line and the whole `do` expression will evaluate to the single value.

Of course, such code won't raise an error:
```javascript
let foo = do {
  if (a) {
    0;
  }  else if (b) {
    1; 
  }  else if (c) {
    2;
  }  else {
    3;
  }
};
```

But such code looks much better:
```javascript
let foo = do {
  if (a) 0;
  else if (b) 1;
  else if (c) 2;
  else 3;
};
```

Not because it saves few lines, but because it just looks closer to pattern matching. So yes, I would still like to fix this bug with `no-unused-expressions`

**Comment by pascalduez** [comment]
I also noticed the plugin erroring on variants that seems fine (similar to early returns), at least it's properly compiled by babel.

```js
let test = do {
 if (a) {
  'a';
 }
 if (b) {
  'b';
 }
 'c';
}
````

**Comment by ljharb** [comment]
This still seems like an issue.

**Comment by devinrhode2** [comment]
@PinkaminaDianePie are semi-colons required in this example of yours
```
let foo = do {
  if (a) 0;
  else if (b) 1;
  else if (c) 2;
  else 3;
};
```

**Comment by PinkaminaDianePie** [comment]
@devinrhode2 honestly I don't remember, I gave up on do-expressions because they are not supported by type systems and even considered harmful for code quality: https://github.com/tc39/proposal-do-expressions/issues/34

**Comment by kaicataldo** [comment]
Thank you for the issue. Now that [@babel/eslint-plugin has been released](https://babeljs.io/blog/2020/07/13/the-state-of-babel-eslint), we are making this repository read-only. If this is a change you would still like to advocate for, please reopen this in the [babel/babel monorepo](https://github.com/babel/babel).
